### PR TITLE
DENA-671: validate replication factor

### PR DESCRIPTION
**Pull request by sbuliarca**
Started the topic configuration rule.
- **Implemented the validation for the replication_factor attribute.**

Will add more on top of it, but want to release it gradually, as every check that we can skip on PRs is a plus for us

### CODE REVIEW

File: rules/msk_topic_config.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	if err := r.validateReplicationFactor(runner, topic); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const (
+	replFactorAttrName   = "replication_factor"
+	replicationFactorVal = 3
```
Comment (matthewhughes-uw) [inline rules/msk_topic_config.go:88]: could you link to some docs explaining why this value

File: rules/msk_topic_config.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		return nil
+	}
+
+	var replFactor int
+	diags := gohcl.DecodeExpression(replFactorAttr.Expr, nil, &replFactor)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	if replFactor != 3 {
```
Comment (matthewhughes-uw) [inline rules/msk_topic_config.go:130]:
```suggestion
	if replFactor != replicationFactorVal {
```

File: rules/msk_topic_config_test.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}{
+		{
+			name: "topic without name",
+			input: `
+resource "kafka_topic" "topic_without_name" {
+}`,
+			expected: []*helper.Issue{
+				{
+					Rule:    rule,
+					Message: "missing replication_factor: it must be equal to '3'",
```
Comment (matthewhughes-uw) [inline rules/msk_topic_config_test.go:30]: should this alert about the missing name, not the replication factor?

Comment (sbuliarca) [reply rules/msk_topic_config_test.go:30]: That's the thing, as the name is handled in the msk_topic_name. 
The thing is that if there is no name I couldn't find an elegant way to propose a fix (inserting the text in the block), thus we only emit the issue.

File: rules/msk_topic_config.go (hunk 1 above)
Comment (sbuliarca) [reply rules/msk_topic_config.go:89]: Added in https://github.com/utilitywarehouse/tflint-ruleset-kafka-config/pull/12/commits/f78223246610da3b0c599389c05720d470de5b0a

File: rules/msk_topic_config_test.go (hunk 3 above)
Comment (matthewhughes-uw) [reply rules/msk_topic_config_test.go:30]: > The name is handled in the msk_topic_name rule.

Can we just skip checking it here then and let that catch it?

>  The thing is that if there is no name I couldn't find an elegant way to propose a fix (inserting the text in the block), thus we only emit the issue.

But the emitted issue feels confusing: the plugin couldn't find a topic name, but spits out an error that the replication factor is wrong/missing

Comment (sbuliarca) [reply rules/msk_topic_config_test.go:30]: So the logic in this plugin is:
if the replication factor is missing:
    if the name is present emit issue with suggested fix to put the replication factor after the name
    if the name is not present emit only the issue without a fix

>But the emitted issue feels confusing: the plugin couldn't find a topic name, but spits out an error that the replication factor is wrong/missing

So we're emitting the issue only when the replication factor is missing. It is not related to the topic name. We're checking here that if no topic name, no fixes are proposed, but the issue is still emited.

Comment (matthewhughes-uw) [reply rules/msk_topic_config_test.go:30]: > So we're emitting the issue only when the replication factor is missing

This is emitted in the `if !hasName` block?

Comment (sbuliarca) [reply rules/msk_topic_config_test.go:30]: Yes, it is this part: https://github.com/utilitywarehouse/tflint-ruleset-kafka-config/blob/topic-configs/rules/msk_topic_config.go#L99-L108

Comment (sbuliarca) [reply rules/msk_topic_config_test.go:30]: Updated the scenario names in here: https://github.com/utilitywarehouse/tflint-ruleset-kafka-config/pull/12/commits/f86c05e33649fc5d8210dc7887b18a8bde3b2c75

Comment (matthewhughes-uw) [reply rules/msk_topic_config_test.go:30]: Ahhh I see now, my reading skills failed and I missed the `hasReplFactor` check just above that :facepalm:

File: rules/msk_topic_config.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+			fmt.Sprintf("missing replication_factor: it must be equal to '%d'", replicationFactorVal),
+			topic.DefRange,
+			func(f tflint.Fixer) error {
+				return f.InsertTextAfter(nameAttr.Range, "\n"+replFactorFix)
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("emitting issue with fix: no replication factor: %w", err)
+		}
+		return nil
```
Comment (matthewhughes-uw) [inline rules/msk_topic_config.go:122]: I think this function could be a bit simpler if this is extracted to a separate function

```suggestion
		r.reportMissingReplicationFactor(runner, topic)
```

**Review by matthewhughes-uw — APPROVED** [review]

File: rules/msk_topic_config.go (hunk 4 above)
Comment (sbuliarca) [reply rules/msk_topic_config.go]: Yep, sounds good. Done in https://github.com/utilitywarehouse/tflint-ruleset-kafka-config/pull/12/commits/03e2e21328b278a42e45133b94b10d29f95b61e9

**Pull request merged**